content/doc/developer/plugin-development/incrementals.adoc

```diff
@@ -0,0 +1,108 @@
+---
+layout: developersection
+title: "Developing plugins and Jenkins core parallely"
+tags:
+- jenkins
+- incrementals
+- core
+- plugin
+description: "This article explains how to use Incrementals for developing plugins parallely with Jenkins core"
+---
+
+A common problem one might stumble upon is developing an API in Jenkins core and simultaneously
+building a reference implementation inside a plugin.
+What to do, one might ask.
+
+Worry not, Incrementals are here to the rescue!
+
+With incrementals, we can create Jenkins core incremental versions which can be used along with the plugin to build the
+plugin against the very latest changes you have made to core.
+Also, it allows keeping track and switching between incremental versions.
+Not only that, it allows others to also not have to build Jenkins locally to try out the changes or collaborate with
+you on your awesome project!
+
+Sounds cool? Let's see how to go about it.
+
+== 1. Create a Pull Request in Jenkins core
+
+First step is to create a PR on github link:https://github.com/jenkinsci/jenkins/pulls[here] by clicking the "New Pull
+Request" button and push the changes to Jenkins core.
+
+image::/images/developer/plugin-development/incrementals/1.png[]
+
+Note: Make sure that you rebase your branch with the latest changes from master, otherwise incrementals version
+will NOT be created.
+
+== 2. Wait for All checks to pass
+
+Wait for 'continuous-integration/jenkins' to perform checks, this can take about two hours.
+
+== 3. Obtain the Incrementals version
+
+After the checks are completed, you should get a message like below:
+
+image::/images/developer/plugin-development/incrementals/3.png[]
+
+Click on 'Details' adjacent to 'continuous-integration/jenkins/incrementals — Deployed to Incrementals.'
+
+Note: If you only see one check 'continuous-integration/jenkins/pr-merge — This commit looks good' then there is an
+error due to which incrementals version was not created. One of the most likely reasons is that latest changes from
+master have not been pulled into the branch. You can click on 'Details' adjacent to it to investigate this error from
+the logs.
+
+== 4. Copy the Version ID
+
+You should be redirected to a page like below:
+
+image::/images/developer/plugin-development/incrementals/4.png[]
+
+Copy the version ID, which in this example would be `2.238-rc29961.5c3c5871cca6`.
+
+== 5. Update the plugin POM
+
+There are two changes you want to make to your plugin POM.
+
+First, update the Jenkins version to the version ID copied in the previous step:
+
+```
+...
+    <properties>
+        <jenkins.version>2.238-rc29961.5c3c5871cca6</jenkins.version>
+        ...
+    </properties>
+...
+```
+
+Then, add the incrementals repository to the list of repositories:
```

Comment:
isn't this handled by the parent pom? / the incrementals plugin?

Reply:
Hi @timja! It wasn't handled automatically when I tried it, which eventually led me to write this guide :P

Reply:
from what I understand it should be handled if the plugin is incrementalised  https://github.com/jenkinsci/incrementals-tools

but it doesn't work well for me locally in intellij (issues with dependency loading IIRC), I use SNAPSHOTs and then swap to incrementals for the CI build, but maybe this would work better

cc @jglick

Reply:
@timja I had a question actually (apologies if it is very simple) isn't this link for plugins so the plugins can create incrementals versions? In this particular guide I have not aimed at creating incrementals for the plugin, but for the Jenkins core version to use against.

Reply:
isn't it the same steps? or does it not work right?

Reply:
Personally, it did not work for me, i got an error ` Unexpected tag: plugin-3.50`.

Reply:
Do not do this! Having `-Pconsume-incrementals` in `.mvn/maven.config` (done for you by `mvn incrementals:incrementalify`) is all you need.

If IDEA does not honor this flag, that is a bug in your IDE, which you may be able to work around by manually configuring the IDEA project to use this profile. NetBeans handles this as of https://github.com/apache/netbeans/pull/1209 FWIW.